ibtk/src/solvers/impls/PETScLevelSolver.cpp

```diff
@@ -307,6 +314,19 @@ void PETScLevelSolver::deallocateSolverState()
     return;
 } // deallocateSolverState
 
+void PETScLevelSolver::addLinearOperator(Mat& op, MatStructure nonzero_pattern)
+{
+#if !defined(NDEBUG)
+	TBOX_ASSERT(!d_is_initialized);
+	TBOX_ASSERT(op);
+#endif
+
+	d_petsc_extern_mat = op;
+	d_extern_mat_nz_pattern = nonzero_pattern;
+
+	return;
+}// addLinearOperator
+
 /////////////////////////////// PROTECTED ////////////////////////////////////
 
 void PETScLevelSolver::init(Pointer<Database> input_db, const std::string& default_options_prefix)
```

Comment:
We (maybe it's just me) are thinking (guessing) that for PCASM to work properly, there is an explicit need to add two matrices, Mat A and Mat B, which have different nonzero patterns, via MATAXPY. If that is not a requirement, we can use MATSHELL. 

P.S: In the application code, A is the Stokes operator and B is the elasticity operator that takes fluid velocity and generates elastic force. B's nonzero pattern is a function of Lagrangian mesh etc...